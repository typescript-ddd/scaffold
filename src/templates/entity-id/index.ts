export * from "./entity-id.generator";
export * from "./entity-id.template";
export * from "./entity-id.types";
