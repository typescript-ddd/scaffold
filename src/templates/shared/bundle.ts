import { GenerateContext, Chunk } from "./generate-context";
import { Template, TemplateOption } from "./template";

export interface Bundle<
  TValues extends { [K in keyof TOptions]: any },
  TOptions extends { [key: string]: TemplateOption }
> extends Omit<Template<TValues, TOptions>, "generate"> {
  generate(value: TValues, context: GenerateContext): Chunk[];
}
