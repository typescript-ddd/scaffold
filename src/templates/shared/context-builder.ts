import { GenerateContext, ContextImportDeclaration } from "./generate-context";
import { arrays } from "../../utils";
import { SourceFile } from "ts-morph";

export class ContextBuilder implements GenerateContext {
  private _fileName: string;
  private _importDeclarations: ContextImportDeclaration[] = [];

  private constructor(
    private rootDir: string = "@/src",
    fileName: string = "example.ts"
  ) {
    this._fileName = fileName;
  }

  resolveDir(
    type: "domain" | "infrastructure" | "application" | "presentation",
    subPath?: string
  ): string {
    switch (type) {
      case "domain":
        return `${this.rootDir}/domain${subPath ? `/${subPath}` : ""}`;
      case "infrastructure":
        return `${this.rootDir}/infrastructure${subPath ? `/${subPath}` : ""}`;
      case "application":
        return `${this.rootDir}/application${subPath ? `/${subPath}` : ""}`;
      case "presentation":
        return `${this.rootDir}/presentation${subPath ? `/${subPath}` : ""}`;
      default:
        throw new Error(`Unknown type ${type} for resolveDir`);
    }
  }

  getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[] {
    const modules = this.getModuleSpecifiers();

    const declarations: ContextImportDeclaration[] = [];

    for (const currentModule of modules) {
      const namedImports = this.getNamedImports(currentModule);
      declarations.push({ moduleSpecifier: currentModule, namedImports });
    }

    if (moduleSpecifier) {
      return declarations.filter((x) => x.moduleSpecifier === moduleSpecifier);
    }

    return declarations;
  }

  addImportDeclaration(declaration: ContextImportDeclaration): void {
    this._importDeclarations.push(declaration);
  }

  getModuleSpecifiers(): string[] {
    return arrays.distinct(
      this._importDeclarations.map((x) => x.moduleSpecifier)
    );
  }

  getNamedImports(moduleSpecifier: string): string[] {
    return arrays.distinct(
      this._importDeclarations
        .filter((x) => x.moduleSpecifier === moduleSpecifier)
        .map((x) => x.namedImports)
        .flat()
    );
  }

  writeSourceFileImports(sourceFile: SourceFile) {
    const importDeclarations = this.getImportDeclarations();
    const moduleSpecifiers = this.getModuleSpecifiers();

    const coreImports = importDeclarations.filter(
      (x) => x.moduleSpecifier === "@typescript-ddd/core"
    );

    if (coreImports) {
      sourceFile.addImportDeclaration({
        moduleSpecifier: `@typescript-ddd/core`,
        namedImports: coreImports.map((x) => x.namedImports).flat(),
      });
    }

    // Non core and root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter(
        (x) => x !== "@typescript-ddd/core" && !x.startsWith(this.rootDir)
      )
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }

    // Root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter((x) => x.startsWith(this.rootDir))
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }
  }

  clearImportDeclarations(): void {
    this._importDeclarations = [];
  }

  tokenize(text: string, tokens: { name: string }[]): string {
    let output = text;
    tokens.forEach((token) => {
      output = output.replace(`{{${token.name}}}`, "${" + token.name + "}");
    });
    return "`" + output + "`";
  }

  get fileName(): string {
    return this._fileName;
  }

  static build(rootDir: string): ContextBuilder {
    return new ContextBuilder(rootDir);
  }
}
