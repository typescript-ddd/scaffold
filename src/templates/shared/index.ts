export * from "./context-builder";
export * from "./generate-context";