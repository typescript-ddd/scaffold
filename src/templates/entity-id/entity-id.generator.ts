import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";
import { EntityIdTemplateValues } from "./entity-id.types";

export const generateEntityId = (
  values: EntityIdTemplateValues,
  context: GenerateContext
) => {
  const { entityName } = values;
  const name = strings.capitalize(entityName);
  const idName = `${name}Id`;
  const fileName = strings.kebab(entityName);

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${fileName}.ts`,
    `
export class ${idName} extends EntityId {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["EntityId", "Uuid"],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(idName);
  classDeclaration.addJsDoc({
    description: `Represents the unique identifier of ${strings.describe(
      name
    )}.`,
    tags: [{ tagName: "extends", text: "{EntityId}" }],
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: [{ name: "value", type: "string" }],
  });
  constructorDeclaration.setBodyText(`super(value);`);
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      name
    )} identifier.`,
    tags: [{ tagName: "param", text: "value The value of the identifier." }],
  });

  classDeclaration
    .addMethod({
      name: "create",
      isStatic: true,
      returnType: idName,
    })
    .setBodyText(`return new ${idName}(Uuid.create().value);`)
    .insertJsDoc(0, {
      description: `Creates a new instance of ${strings.describe(
        name
      )} identifier.`,
      tags: [
        {
          tagName: "returns",
          text: `{${idName}} - ${strings.capitalize(
            strings.describe(name)
          )} identifier.`,
        },
      ],
    });

  classDeclaration
    .addMethod({
      name: "fromValue",
      isStatic: true,
      parameters: [{ name: "value", type: "string" }],
      returnType: idName,
    })
    .setBodyText(`return new ${idName}(value);`)
    .insertJsDoc(0, {
      description: `Creates a new instance of ${strings.describe(
        name
      )} identifier from its value.`,
      tags: [
        {
          tagName: "param",
          text: "{string} value - The value of the identifier.",
        },
        {
          tagName: "throws",
          text: `{InvariantViolationError} - When the value is null, empty or invalid.`,
        },
        {
          tagName: "returns",
          text: `{${idName}} - ${strings.capitalize(
            strings.describe(name)
          )} identifier.`,
        },
      ],
    });

  let output = sourceFile.getFullText();

  return output;
};
