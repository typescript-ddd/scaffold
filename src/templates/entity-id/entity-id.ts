import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateEntityIdOptions = {
  entityName: string;
};

export const generateEntityId = (
  options: GenerateEntityIdOptions,
  context: GenerateContext
) => {
  const { entityName } = options;
  const name = strings.capitalize(entityName);
  const idName = `${name}Id`;
  const fileName = strings.kebab(entityName);

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${fileName}.ts`,
    `
export class ${idName} extends EntityId {}    
`
  );

  sourceFile.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["EntityId", "Uuid"],
  });

  const classDeclaration = sourceFile.getClass(idName);
  classDeclaration.addJsDoc({
    description: `Represents the unique identifier of ${strings.describe(
      name
    )}.`,
    tags: [{ tagName: "extends", text: "{EntityId}" }],
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: [{ name: "value", type: "string" }],
  });
  constructorDeclaration.setBodyText(`super(value);`);
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      name
    )} identifier.`,
    tags: [{ tagName: "param", text: "value The value of the identifier." }],
  });

  classDeclaration
    .addMethod({
      name: "create",
      isStatic: true,
      returnType: idName,
    })
    .setBodyText(`return new ${idName}(Uuid.create().value);`)
    .insertJsDoc(0, {
      description: `Creates a new instance of ${strings.describe(
        name
      )} identifier.`,
      tags: [
        {
          tagName: "returns",
          text: `{${idName}} - ${strings.capitalize(strings.describe(name))} identifier.`,
        },
      ],
    });

  classDeclaration
    .addMethod({
      name: "fromValue",
      isStatic: true,
      parameters: [{ name: "value", type: "string" }],
      returnType: idName,
    })
    .setBodyText(`return new ${idName}(value);`)
    .insertJsDoc(0, {
      description: `Creates a new instance of ${strings.describe(
        name
      )} identifier from its value.`,
      tags: [
        {
          tagName: "param",
          text: "{string} value - The value of the identifier.",
        },
        {
            tagName: "throws",
            text: `{InvariantViolationError} - When the value is null, empty or invalid.`,
        },
        {
          tagName: "returns",
          text: `{${idName}} - ${strings.capitalize(strings.describe(name))} identifier.`,
        }
      ],
    });

  let output = sourceFile.getFullText();

  return output;
};
