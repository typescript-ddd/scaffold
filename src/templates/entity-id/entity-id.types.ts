import { TemplateStringOption } from "../shared/template";

export type EntityIdTemplateValues = {
  entityName: string;
};

export type EntityIdTemplateOptions = {
  entityName: TemplateStringOption;
};
