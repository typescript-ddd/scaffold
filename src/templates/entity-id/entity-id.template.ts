import { GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateEntityId } from "./entity-id.generator";
import {
  EntityIdTemplateOptions,
  EntityIdTemplateValues,
} from "./entity-id.types";

export class EntityIdTemplate
  implements
    Template<EntityIdTemplateValues, EntityIdTemplateOptions>
{
  readonly name = "Entity ID";
  readonly description = "Generates an entity ID";
  readonly options: EntityIdTemplateOptions = {
    entityName: {
      name: "entityName",
      label: "Entity Name",
      description: "The name of the entity",
      optionType: "string",
    },
  };
  readonly defaultValues = {
    entityName: "User",
  };
  generate(values: EntityIdTemplateValues, context: GenerateContext): string {
    return generateEntityId(values, context);
  }
}
