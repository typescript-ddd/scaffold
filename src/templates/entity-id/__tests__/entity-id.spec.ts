import { GenerateContext } from "../../shared";
import { ContextBuilder } from "../../shared/context-builder";
import { generateEntityId } from "../entity-id";

describe("EntityId", () => {
  let context: GenerateContext;

  beforeEach(() => {
    context = ContextBuilder.build("@src");
  });

  it("Should render the entity id component", () => {
    const output = generateEntityId({ entityName: "User" }, context);

    expect(output).toMatchSnapshot();
  });
});
