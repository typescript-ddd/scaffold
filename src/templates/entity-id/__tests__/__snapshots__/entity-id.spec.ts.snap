// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityId Should render the entity id component 1`] = `
"import { EntityId, Uuid } from "@typescript-ddd/core";

/**
 * Represents the unique identifier of a user.
 * @extends {EntityId}
 */
export class UserId extends EntityId {
    /**
     * Initializes a new instance of a user identifier.
     * @param value The value of the identifier.
     */
    private constructor(value: string) {
        super(value);
    }

    /**
     * Creates a new instance of a user identifier.
     * @returns {UserId} - A user identifier.
     */
    static create(): UserId {
        return new UserId(Uuid.create().value);
    }

    /**
     * Creates a new instance of a user identifier from its value.
     * @param {string} value - The value of the identifier.
     * @throws {InvariantViolationError} - When the value is null, empty or invalid.
     * @returns {UserId} - A user identifier.
     */
    static fromValue(value: string): UserId {
        return new UserId(value);
    }
}    
"
`;
